### PR TITLE
[core] index mapping push down to format reader and skip partition value reading from file

### PR DESCRIPTION
**Pull request by leaves12138**
### Purpose

1】 index mapping use ProjectedRow has a pool performance, use index mapping push down to format and expand vectors directly could speed it up

2】 partition value is not needed while reading from file, we skip reading partition columns from files

### API and Format

No influence to user if there's no bug

### CODE REVIEW

File: paimon-format/src/main/java/org/apache/paimon/format/orc/OrcReaderFactory.java [hunk 1]
```diff
@@ -93,9 +95,28 @@ public OrcVectorizedReader createReader(FileIO fileIO, Path file) throws IOExcep
     }
 
     @Override
-    public OrcVectorizedReader createReader(FileIO fileIO, Path file, int poolSize)
+    public OrcVectorizedReader createReader(FileIO fileIO, Path file, int pooSize)
```
Comment (JingsongLi) [inline paimon-format/src/main/java/org/apache/paimon/format/orc/OrcReaderFactory.java:98]: revert?

File: paimon-common/src/main/java/org/apache/paimon/format/FormatReaderFactory.java [hunk 2]
```diff
@@ -33,4 +36,19 @@ public interface FormatReaderFactory extends Serializable {
 
     RecordReader<InternalRow> createReader(FileIO fileIO, Path file, int poolSize)
             throws IOException;
+
+    RecordReader<InternalRow> createReader(
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/format/FormatReaderFactory.java:40]: Do we need to introduce these methods?

I think we can just handle return value from `RecordReader`.
```
if (it is a partitioned table) {
  if (it is a `ColumnarRowIterator`) {
    then use `VectorMappingUtils.wrap`;
  } else {
    use a PartitionedRow to wrap it;
  }
}
```

**Review by JingsongLi — CHANGES_REQUESTED** [review]

File: paimon-core/src/main/java/org/apache/paimon/operation/AppendOnlyFileStoreRead.java [hunk 3]
```diff
@@ -112,41 +115,65 @@ public RecordReader<InternalRow> createReader(DataSplit split) throws IOExceptio
...
+
                                 return new BulkFormatMapping(
                                         indexCastMapping.getIndexMapping(),
                                         indexCastMapping.getCastMapping(),
+                                        partitionPair,
                                         formatDiscover
                                                 .discover(formatIdentifier)
                                                 .createReaderFactory(
-                                                        rowType, dataProjection, dataFilters));
+                                                        dataSchema.logicalRowType(),
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/operation/AppendOnlyFileStoreRead.java:161]: The type still contains Partition fields? It is weird.

Can you just modify to `FormatReaderFactory createReaderFactory(RowType projectedType, @Nullable List<Predicate> filters)`?

Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/operation/AppendOnlyFileStoreRead.java]: Done this.

File: paimon-common/src/main/java/org/apache/paimon/PartitionSettedRow.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+import org.apache.paimon.data.InternalMap;
+import org.apache.paimon.data.InternalRow;
+import org.apache.paimon.data.PartitionInfo;
+import org.apache.paimon.data.Timestamp;
+import org.apache.paimon.types.RowKind;
+
+/** An implementation of {@link InternalRow} which provides a row the fixed partition value. */
+public class PartitionSettedRow implements InternalRow {
+
+    private final BinaryRow paritition;
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/PartitionSettedRow.java:34]: partition

File: paimon-common/src/main/java/org/apache/paimon/data/columnar/ColumnarRow.java [hunk 5]
```diff
@@ -221,4 +225,18 @@ public int hashCode() {
         throw new UnsupportedOperationException(
                 "ColumnarRowData do not support hashCode, please hash fields one by one!");
     }
+
+    public void mapping(@Nullable PartitionInfo partitionInfo, @Nullable int[] indexMapping) {
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/data/columnar/ColumnarRow.java:229]: Return a new `ColumnarRow`.
If there is no performance issue, it is better to keep objects imummable.

File: paimon-common/src/main/java/org/apache/paimon/data/columnar/ColumnarRowIterator.java [hunk 6]
```diff
@@ -55,4 +57,13 @@ public InternalRow next() {
...
         }
     }
+
+    public ColumnarRowIterator mapping(
+            @Nullable PartitionInfo partitionInfo, @Nullable int[] indexMapping) {
+        if (!mapped) {
+            rowData.mapping(partitionInfo, indexMapping);
+            mapped = true;
+        }
+        return this;
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/data/columnar/ColumnarRowIterator.java:67]: Do not return this, return a new `ColumnarRowIterator`.

File: paimon-common/src/main/java/org/apache/paimon/data/columnar/ColumnarRowIterator.java [hunk 7]
```diff
@@ -31,6 +32,7 @@
 public class ColumnarRowIterator extends RecyclableIterator<InternalRow> {
 
     private final ColumnarRow rowData;
+    private boolean mapped = false;
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/data/columnar/ColumnarRowIterator.java:35]: remove this field.

File: paimon-core/src/main/java/org/apache/paimon/io/AbstractFileRecordIterator.java [hunk 8]
```diff
@@ -32,11 +34,17 @@
  * @param <V> the row type.
  */
 public abstract class AbstractFileRecordIterator<V> implements RecordReader.RecordIterator<V> {
+
+    @Nullable private final PartitionSettedRow partitionSettedRow;
     @Nullable private final ProjectedRow projectedRow;
     @Nullable private final CastedRow castedRow;
 
     protected AbstractFileRecordIterator(
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/io/AbstractFileRecordIterator.java:42]: We need to refine this Iterator.
For example, use `RecordIterator.transform`.

File: paimon-format/src/main/java/org/apache/paimon/format/avro/FieldReaderFactory.java [hunk 9]
```diff
@@ -432,49 +482,42 @@ public void skip(Decoder decoder) throws IOException {
...
-                    mapping[projection[i]] = columns;
+            this.mapping = new int[fields.size()];
+            this.mappingBack = new int[schemaFields.size()];
+            Arrays.fill(mappingBack, -1);
+            for (int i = 0; i < mapping.length; i++) {
+                DataField field = fields.get(i);
+                Schema.Field schemaField = schema.getField(field.name());
+                if (schemaField != null) {
+                    int index = schemaFields.indexOf(schemaField);
+                    this.mapping[i] = schemaFields.indexOf(schemaField);
```
Comment (JingsongLi) [inline paimon-format/src/main/java/org/apache/paimon/format/avro/FieldReaderFactory.java:506]: use index

File: paimon-format/src/main/java/org/apache/paimon/format/avro/FieldReaderFactory.java [hunk 10]
```diff
@@ -483,24 +526,27 @@ public RowReader(Schema schema, List<DataType> fieldTypes, int[] projection) {
...
                 } else {
-                    Object value = reader.read(decoder, row.getField(columns[0]));
-                    for (int column : columns) {
-                        row.setField(column, value);
-                    }
+                    fieldReaders[i].skip(decoder);
                 }
             }
+
+            values = VectorMappingUtils.mappingObjects(values, mapping);
```
Comment (JingsongLi) [inline paimon-format/src/main/java/org/apache/paimon/format/avro/FieldReaderFactory.java:544]: It is weird, there is a for loop to values. We don't need this `mappingObjects`.

File: paimon-format/src/main/java/org/apache/paimon/format/avro/FieldReaderFactory.java [hunk 11]
```diff
@@ -120,31 +129,63 @@ public FieldReader visitTimestampMillis(int precision) {
         return TIMESTAMP_MILLS_READER;
     }
 
+    @Override
+    public FieldReader visitTimestampMillis() {
```
Comment (JingsongLi) [inline paimon-format/src/main/java/org/apache/paimon/format/avro/FieldReaderFactory.java:133]: You can add change origin method `visitTimestampMillis(@Nullable Integer precision)`.

**Review by JingsongLi — CHANGES_REQUESTED** [review]

File: paimon-common/src/test/java/org/apache/paimon/utils/VectorMappingUtilsTest.java [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+import org.apache.paimon.types.DataTypes;
+import org.apache.paimon.types.RowType;
+
+import org.assertj.core.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+
+/** Tests for {@link VectorMappingUtils}. */
+public class VectorMappingUtilsTest {
```
Comment (JingsongLi) [inline paimon-common/src/test/java/org/apache/paimon/utils/VectorMappingUtilsTest.java:35]: Add tests for all types.

Comment (leaves12138) [reply paimon-common/src/test/java/org/apache/paimon/utils/VectorMappingUtilsTest.java:45]: Done

File: paimon-format/src/main/java/org/apache/paimon/format/avro/FieldReaderFactory.java (hunk 10 above)
Comment (leaves12138) [reply paimon-format/src/main/java/org/apache/paimon/format/avro/FieldReaderFactory.java]: Fixed this.

File: paimon-format/src/main/java/org/apache/paimon/format/avro/FieldReaderFactory.java (hunk 9 above)
Comment (leaves12138) [reply paimon-format/src/main/java/org/apache/paimon/format/avro/FieldReaderFactory.java]: Fixed this.

File: paimon-core/src/main/java/org/apache/paimon/io/AbstractFileRecordIterator.java (hunk 8 above)
Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/io/AbstractFileRecordIterator.java]: Fixed this and removed class `AbstractFileRecordIterator`

File: paimon-common/src/main/java/org/apache/paimon/data/columnar/ColumnarRowIterator.java (hunk 7 above)
Comment (leaves12138) [reply paimon-common/src/main/java/org/apache/paimon/data/columnar/ColumnarRowIterator.java]: Fixed this.

File: paimon-common/src/main/java/org/apache/paimon/data/columnar/ColumnarRowIterator.java (hunk 6 above)
Comment (leaves12138) [reply paimon-common/src/main/java/org/apache/paimon/data/columnar/ColumnarRowIterator.java:78]: Fixed this.

File: paimon-common/src/main/java/org/apache/paimon/data/columnar/ColumnarRow.java (hunk 5 above)
Comment (leaves12138) [reply paimon-common/src/main/java/org/apache/paimon/data/columnar/ColumnarRow.java]: Fixed this.

File: paimon-common/src/main/java/org/apache/paimon/PartitionSettedRow.java (hunk 4 above)
Comment (leaves12138) [reply paimon-common/src/main/java/org/apache/paimon/PartitionSettedRow.java]: Fixed this.

File: paimon-common/src/main/java/org/apache/paimon/PartitionSettedRow.java [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+
+/** An implementation of {@link InternalRow} which provides a row the fixed partition value. */
+public class PartitionSettedRow implements InternalRow {
+
+    private final BinaryRow partition;
+    private final PartitionInfo partitionInfo;
+
+    protected InternalRow row;
+
+    protected PartitionSettedRow(PartitionInfo partitionInfo) {
```
[the remark this answers is not in the document]
Comment (leaves12138) [reply paimon-common/src/main/java/org/apache/paimon/PartitionSettedRow.java:39]: Fixed this.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**